js/core.js

```diff
@@ -1539,80 +1541,115 @@ var ERMrest = (function (module) {
+                columns = module._getRecursiveAnnotationValue(context, this._table.annotations.get(module._annotations.VISIBLE_COLUMNS).content);
             }
 
-            // get from annotation             
-            if (orders !== -1) {
-                var addedFKs = {}, // to avoid duplicate 
-                    fkName, 
-                    colFound,
-                    addCol;
-
-                for (i = 0; i < orders.length; i++) {
-                    addCol = true;
-                    if (Array.isArray(orders[i])) {
-                        fk = this._table.schema.catalog.constraintByNamePair(orders[i]);
-                        if (fk !== null) {
-                            // check if FK of this table and avoid duplicate
-                            fkName = fk.constraint_names[0].join(":");
-                            if (fk._table != this._table || (fkName in addedFKs) || hideFKR(fk)) {
-                                addCol = false;
-                            } else {
-                                addedFKs[fkName] = 1;
-                                col = new PseudoColumn(fk, fk.colset.columns[0]);
-                            }
-                        } else {
-                            addCol = false;
+            // use columns as the base if annotation was not present
+            if (columns === -1) {
+                columns = this.all();
+            }
+
+            // columns is a list of string (column name), array (fk constraint name) or objects (column)
+            for (i = 0; i < columns.length; i++) {
+                col = columns[i];
+                if (Array.isArray(col)) { // foreign keys
+                    fk = this._table.schema.catalog.constraintByNamePair(col);
+                    if (fk !== null) {
+                        fkName = fk.constraint_names[0].join(":");
+
+                        // fk is in this table, avoid duplicate and it's not hidden.
+                        if (!(fkName in addedFKs) && fk._table == this._table && !hideFKR(fk)) {
+                            addedFKs[fkName] = 1;
+                            visiblePseudoColumns.push(new PseudoColumn(fk, fk.colset.columns[0]));
                         }
-                    } else {
-                        colFound = false;
-                        for (j=0; j < columns.length && !colFound; j++) {
-                            if (columns[j].name == orders[i]) {
-                                col = columns[j];
-                                colFound = true;
+                    }
+                } else { // columns
+
+                    // (annotation: column name)
+                    if (typeof col == "string" && !(col in consideredColumns)) {
+                        // find the corresponding column
+                        for (j = 0; j < this.length(); j++) {
+                            if (this.getByPosition(j).name == col) {
+                                col = this.getByPosition(j);
```

Comment:
Could you have done this with this.get(col) instead of the loop? Just to simplify the code.